### ISSUE TITLE
Assemblies have no strong name

### ISSUE THREAD

**Issue by lwuckel** [LightBDD/LightBDD#157]
My projects are signed with a strong name and my test-projects are defined as friends. 
So the test-projects have to signed too.

The issue is that the LightBDD Assemblies are not signed. So I can't use it.
I get following warning message:

CSC : warning CS8002: Die referenzierte Assembly "LightBDD.NUnit3, Version=2.4.3.0, Culture=neutral, PublicKeyToken=null" besitzt keinen starken Namen.
CSC : warning CS8002: Die referenzierte Assembly "LightBDD.Framework, Version=2.4.3.0, Culture=neutral, PublicKeyToken=null" besitzt keinen starken Namen.

The TestRunner can't find the Test-Methods.
Could you provide a new signed package too?

Or is there annother workaround?

**Comment by Suremaker** [comment]
Hello @lwuckel ,

I checked what is needed to sign the LightBDD project and found that while it is an easy task, it is also a breaking change 😞 
The reason is that it is not possible to have a project that would use mixture of unsigned and signed version of the LightBDD and there is no possibility to use binding redirection or anything else to fix it.

Assuming project structure like:
* TestAbstractions 
  -> LightBDD.NUnit3 2.4.3
* TestProject 
  -> LightBDD.NUnit3 2.5.0
  -> TestAsbtractions

The compilation of `TestProject` would fail with error: 
```
The type 'FeatureFixture' is defined in an assembly that is not referenced. You must add a reference to assembly 'LightBDD.NUnit3, Version=2.4.3.0, Culture=neutral, PublicKeyToken=null'.
```

As all the LightBDD versions up to 2.4.3 are not signed. If I introduce the 2.5.0 as signed one, it would mean that all users of the LightBDD will have to recompile the whole code to migrate to 2.5.0, which would be a pain for the users that built some common/abstraction projects around LightBDD. 

Because of the above, I would prefer not to introduce the strong name for LightBDD at this moment, but defer it for major update of LightBDD (3X series) that will happen in future... if that is possible.

-------

Having said that, I do not want to make it a showstopper for you, so I wonder if there is any workaround you could use.

### The obvious question is if it is possible to make the test projects not being signed in the first place?

You have mentioned that the reason for signing test project are friend assemblies so I guess you are using `InternalsVisibleToAttribute` to expose internals for the tests.

I wonder if there is any way to avoid using it?

I could imagine a few scenarios that could be worked around (of course I do not know the nature of your project so I am not sure if they are applicable)...

#### 1. If you are exposing internal types in order to unit test them, but use LightBDD for more high level tests (that potentially do not use internals), then perhaps it would be enough to separate the **unit tests** from **lightbdd tests** by making separate projects?

In such scenario, unit tests would be signed itself and use internal types, while the LightBDD tests would be in other, unsigned project that would use only public types. 

#### 2. Maybe some of the internals visibility could be relaxed for tests?

Assuming a following class:
```c#
internal class Calculator
{
  internal int Add(int x, int y) => x+y;
}
```

One of the way to test it without making it fully exposed and easy to misuse is to make it inheritable for test version, but protected by default:
```c# 
public class Calculator
{
  protected internal Calculator() { } //so that it cannot be instantiated easily outside of the project
  protected internal int Add(int x, int y) => x+y;
}

//test project
public class TestableCalculator: Calculator
{
  public int Add(int x, int y) => base.Add(x,y);
}
```
For some specific scenarios I was using such approach, but obviously it is not applicable for all cases.

#### 3. Assuming that you do not want to expose such classes at all and you have only few internal types to test, you can try the proxy assembly approach:

* Domain.csproj (signed)
```c#
[assembly: InternalsVisibleTo("TestHelpers")]

public interface ICalculator
{
  int Add(int x, int y);
}

internal class InternalCalculator: ICalculator
{
  public int Add(int x, int y) => x + y;
}
```
* TestHelpers (signed) -> Domain
```c#
public static class Helpers
{
  // expose internal type via it's interface
  public static ICalculator CreateInternalCalculator() => new InternalCalculator();
}
```
* UnitTests (unsigned) -> TestHelpers, Domain
```c#
[TestFixture]
public class InternalCalculatorTests
{
    [Test]
    public void Adding_numbers()
    {
        ICalculator calculator = Helpers.CreateInternalCalculator();
        Assert.AreEqual(5, calculator.Add(3, 2));
    }
}
```

### Make a signed version of LightBDD 

If none of the above workarounds are applicable, you still can either:
* clone the repo and make signed packages to be used internally by your company,
* or sign the dlls manually (https://blogs.msdn.microsoft.com/miah/2008/02/28/how-to-assign-a-strong-name-to-an-unsigned-3rd-party-assembly/) and put in the repo or even sign them on the fly build if you wish to...


Please let me know if any of the suggestion is helpful?

**Comment by lwuckel** [comment]
Hi @Suremaker 

thank you for your answer, but I don't like workarounds. They need much extra time for maintenance.

At nuget.org you can find many packages that are signed or have an additionally signed package.
For instance Nunit.
Or you can search für signed packages at nuget.org
https://www.nuget.org/packages?q=signed

Maybe you could provide an additionally signed nuget package too?

Currently I compiled the LightBDD source code and signed by a new generated key. This package I've put into my local nuget repository.
But everyone that use signed assemblies too have to do these extra steps. That prevent using LightBDD.

Same issue I've got while using mspec. The signed version is older than the non-signed version. So I had to rebuild the package to sign it.

Maybe it is not so complicated to add an additional step in your build scripts to build all packages signed. That would be great,

Regards Lars

**Comment by Suremaker** [comment]
Hi Lars,

Thanks for coming back with the response.
> Maybe you could provide an additionally signed nuget package too?
> ...
> Same issue I've got while using mspec. The signed version is older than the non-signed version. So I had to rebuild the package to sign it.

This is the reason I would prefer to release one set of packages instead of having a set of unsigned and signed ones. Currently, releasing LightBDD is about publishing 9 packages. If I go to provide a separate set of signed ones, the number of maintained packages will raise to 18.  
I think it would be much better to just upgrade LightBDD to version 3.x and make all packages signed.

> Currently I compiled the LightBDD source code and signed by a new generated key. This package I've put into my local nuget repository.
> But everyone that use signed assemblies too have to do these extra steps. That prevent using LightBDD.

It's good you found a way to make it working. 
I see that it requires more effort right now to use LightBDD with signed test projects, however with the reasons mentioned in previous comment I won't be able to provide the signed version immediately.

Having said that, if the current workaround works for you, I can include this change in the LightBDD 3.0 and focus on delivering it around January - unfortunately the major version upgrade will take a bit more time as I want to sort out few other bits in the framework, while I cannot dedicate every day to work on the project.  
If for that time range it is ok for you to use the current workaround, I can promise to focus on that work in the meantime.

Thanks,
Wojtek

**Comment by Suremaker** [comment]
Implemented in LightBDD 3.0.
